src/cached_object_store/lru.rs

```diff
@@ -0,0 +1,502 @@
+                cache.detach(&mut old_node);
+                let LruEntry { key, value, .. } = old_node;
+                let (k, v) = unsafe { (key.assume_init(), value.assume_init()) };
+                let weight = (cache.weighter)(&k, &v);
+                cache.usage -= weight;
+                Some((k, v))
+            }
+        }
+    }
+}
+
+pub struct LruCache<K, V> {
+    table: HashMap<KeyRef<K>, NonNull<LruEntry<K, V>>>,
+    capacity: NonZeroUsize,
+    weighter: Box<dyn Weighter<K, V>>,
+    usage: usize,
+
+    // Sentinel nodes
+    head: *mut LruEntry<K, V>,
+    tail: *mut LruEntry<K, V>,
+}
+
+impl<K: Eq + Hash, V> LruCache<K, V> {
+    fn new(capacity: NonZeroUsize, weighter: impl Weighter<K, V>) -> Self {
+        let cache = LruCache::<K, V> {
+            table: HashMap::default(),
+            capacity,
+            weighter: Box::new(weighter),
+            usage: 0,
+
+            head: Box::into_raw(Box::new(LruEntry::new_empty())),
+            tail: Box::into_raw(Box::new(LruEntry::new_empty())),
+        };
+
+        unsafe {
+            (*cache.head).next = cache.tail;
+            (*cache.tail).prev = cache.head;
+        }
+
+        cache
+    }
+
+    pub fn len(&self) -> usize {
+        self.table.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[inline]
+    pub fn adjust_capacity(&mut self, new_capacity: NonZeroUsize) {
+        self.capacity = new_capacity;
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> NonZeroUsize {
+        self.capacity
+    }
+
+    fn detach(&mut self, node: *mut LruEntry<K, V>) {
+        unsafe {
+            (*(*node).prev).next = (*node).next;
+            (*(*node).next).prev = (*node).prev;
+        }
+    }
+
+    // Attach node after the head
+    fn attach(&mut self, node: *mut LruEntry<K, V>) {
+        unsafe {
+            (*node).next = (*self.head).next;
+            (*node).prev = self.head;
+            (*self.head).next = node;
+            (*(*node).next).prev = node;
+        }
+    }
+
+    pub fn scope<'cache, F, R>(&'cache mut self, func: F) -> R
+    where
+        for<'brand> F: FnOnce(CacheHandle<'cache, 'brand, K, V>, ValuePerm<'brand>) -> R,
+    {
+        let handle = CacheHandle {
+            _lifetime: Default::default(),
+            cache: self,
+        };
+        let perm = ValuePerm {
+            _lifetime: InvariantLifetime::default(),
+        };
+        func(handle, perm)
+    }
+
+    // Possible evict multiple entries in the LRU to create space for the element to be inserted
+    fn try_evict(&mut self, weight: usize) {
+        while self.usage + weight > self.capacity.get() {
+            // We place the oldest entry at the tail of the list
+            let old_key = KeyRef {
+                key: unsafe { &(*(*(*self.tail).prev).key.as_ptr()) },
+            };
+            let node_ptr = self.table.remove(&old_key).unwrap();
```

Comment:
what's guaranteeing that `unwrap` is safe here?

Reply:
First, I got this from [lru-rs](https://github.com/jeromefroe/lru-rs/blob/master/src/lib.rs#L393).
Second, we will insert `self.table` during put operation.